platform/openide.util/test/unit/src/org/openide/util/PairTest.java

```diff
@@ -19,39 +19,84 @@
-        assertFalse(p1a.equals(p5));
-        assertFalse(p1a.equals(p6));
-        assertEquals(p1a.hashCode(), p1b.hashCode());
-        assertEquals(p4.hashCode(), p4.hashCode());
-        assertEquals(p5.hashCode(), p5.hashCode());
-        assertEquals(p6.hashCode(), p6.hashCode());
-        assertFalse(p4.hashCode() == p5.hashCode());
+        assertTrue(pair11a.equals(pair11a));
+        assertEquals(pair11a.hashCode(), pair11a.hashCode());
+        assertEquals(new Integer(1), pair11a.first());
+        assertEquals(new Integer(1), pair11a.second());
+        
+        assertTrue(pair12.equals(pair12));
+        assertEquals(pair12.hashCode(), pair12.hashCode());
+        assertEquals(new Integer(1), pair12.first());
+        assertEquals(new Integer(2), pair12.second());
+
+        assertTrue(pair11a.equals(pair11b));
+        assertEquals(pair11a.hashCode(), pair11b.hashCode());
+        assertEquals(new Integer(1), pair11b.first());
+        assertEquals(new Integer(1), pair11b.second());
+
+        assertTrue(pair1Null.equals(pair1Null));
+        assertEquals(pair1Null.hashCode(), pair1Null.hashCode());
+        assertEquals(new Integer(1), pair1Null.first());
+        assertEquals(null, pair1Null.second());
+
+        assertTrue(pairNul1.equals(pairNul1));
+        assertEquals(pairNul1.hashCode(), pairNul1.hashCode());
+        assertEquals(null, pairNul1.first());
+        assertEquals(new Integer(1), pairNul1.second());
+
+        assertTrue(pairNullNull.equals(pairNullNull));
+        assertEquals(pairNullNull.hashCode(), pairNullNull.hashCode());
+        assertEquals(null, pairNullNull.first());
+        assertEquals(null, pairNullNull.second());
     }
 
+    //-------------------------------------------------------------------------- 
+    public void test_pairInequality() {
+
+        assertFalse(pair11a.equals(pair12));
+        assertFalse(pair11a.equals(pair21));
+        assertFalse(pair11a.equals(pairNul1));
+        assertFalse(pair11a.equals(pair1Null));
+        assertFalse(pair11a.equals(pairNullNull));
+        assertFalse(pairNul1.equals(pair11a));
+        assertFalse(pair1Null.equals(pair11a));
+        assertFalse(pairNullNull.equals(pair11a));
+        
+        assertFalse(pair11a.equals(null));
+        assertFalse(pair11a.equals(""));
+    }
+    //--------------------------------------------------------------------------
+    public void test_pair_toString() {
+        
+        assertEquals("Pair[1,1]", pair11a.toString());
+        assertEquals("Pair[1,1]", pair11b.toString());
+        assertEquals("Pair[1,2]", pair12.toString());
+        assertEquals("Pair[1,null]", pair1Null.toString());
+        assertEquals("Pair[null,1]", pairNul1.toString());
+        assertEquals("Pair[null,null]", pairNullNull.toString());
+    }
+    //-------------------------------------------------------------------------- 
+    private final Pair<Integer, Integer> pair11a = Pair.of(1, 1);
```

Comment:
If we're testing "Pair" here, we should instantiate Pairs in test methods/setup methods, not as class members. What happens if the constructors themselves fail?

Reply:
good point. I'll fix it